src/hypofuzz/dashboard/dashboard.py

```diff
@@ -204,66 +185,102 @@ def __init__(self, websocket: WebSocket, nodeid: str) -> None:
-            "nodeid": test.nodeid,
-            "failure": test.failure,
+        test_data: AddTestsEvent = {
+            "type": DashboardEventType.ADD_TESTS,
+            "tests": [
+                {
+                    "database_key": test.database_key,
+                    "nodeid": test.nodeid,
+                    "failure": test.failure,
+                }
+            ],
         }
-        await self.send_event({"type": DashboardEventType.ADD_TESTS}, [test_data])
+        await self.send_event(test_data)
 
         # then its reports. Note we don't currently downsample with _sample_reports
         # on individual test pages, unlike the overview page.
         for worker_uuid, reports in test.reports_by_worker.items():
-            report_data = {
+            report_event: AddReportsEvent = {
+                "type": DashboardEventType.ADD_REPORTS,
                 "nodeid": test.nodeid,
                 "worker_uuid": worker_uuid,
-                "reports": [report_for_websocket(report) for report in reports],
+                "reports": [dashboard_report(report) for report in reports],
             }
-            await self.send_event({"type": DashboardEventType.ADD_REPORTS}, report_data)
-
-        await self.send_event(
-            {"type": DashboardEventType.ADD_ROLLING_OBSERVATIONS},
-            {"nodeid": self.nodeid, "observations": test.rolling_observations},
-        )
-        await self.send_event(
-            {"type": DashboardEventType.ADD_CORPUS_OBSERVATIONS},
-            {"nodeid": self.nodeid, "observations": test.corpus_observations},
-        )
+            await self.send_event(report_event)
+
+        # then its observations, sending the observations (without their repr)
+        # first, followed by the repr of each observation.
+        for obs_type, observations in [
+            ("rolling", test.rolling_observations),
+            ("corpus", test.corpus_observations),
+        ]:
+            await self.send_event(
+                {
+                    "type": DashboardEventType.ADD_OBSERVATIONS,
+                    "nodeid": self.nodeid,
+                    "observation_type": obs_type,  # type: ignore
+                    "observations": [
+                        dashboard_observation(obs) for obs in observations
+                    ],
+                },
+            )
+
+        for obs_type, observations in [
+            ("rolling", test.rolling_observations),
+            ("corpus", test.corpus_observations),
+        ]:
+            await self.send_event(
+                {
+                    "type": DashboardEventType.SET_OBSERVATION_REPRS,
+                    "nodeid": self.nodeid,
+                    "observation_type": obs_type,  # type: ignore
+                    "representations": [
+                        {
+                            # using run_start as the observation primary key
+                            "run_start": obs.run_start,
```

Comment:
Makes me nervous; [some clocks have pretty low resolution](https://peps.python.org/pep-0564/#annex-clocks-resolution-in-python).  Seems safe enough if we add the worker_uuid though.